### DISCUSSION TITLE
Lazy Methods

### DISCUSSION THREAD

**Discussion by mezafits** [dotnet/csharplang#6392] [Language Ideas]
Lazy Methods could be a syntax sugar that allows you to configure the execution of a method in a way that would allow a diffrent method to be called before or after another method. 

For Example. 
```



public class ExternalServiceWrapper
{

private GenericServiceClient client;

public ExternalServiceWrapper()
{

}

Lazy void ConnectToExternalService()
{
  /// Some code executes and creates a connection.
  client = new GenericServiceClient("connection info");
  client.connect();
}

 ///You then mark other methods in your class with the Lazy Attribute
 
[Lazy(ConnectToExternalService), OneTime]
List<SomeData> GetDataFromService()
{
 /// this method uses the connection made in the established in the ConnectToExternalService() method
}

[Lazy(ConnectToExternalSerivce), OneTime]
void PostUpdateToService()
{

 /// this method uses the connection made in the established in the ConnectToExternalService() method
 /// Posting data 

}

}

```

For me at least this would be preferable to manually calling a startup method or a init method before usage. I

I have not defined a standard for this keyword and it's usage with methods, but I think there are a number of very practical examples

You could imagine the Lazy attribute could allow for ExecuteOnStart or ExecuteOnComplete, perhaps ExecuteOneTime or ExecuteEachTime ?

**Comment by HaloFour** [comment]
This feels like it falls under the `replace`/`original` proposal for supporting AOP in source generators, which is something that the team has seemingly lost interest in implementing.  #107

**Comment by Xyncgas** [comment]
Coming from F# it's an useful idea, certainly a handy thing there
